### PR TITLE
[MAINT] switch bidsschematools to pyproject.toml

### PR DESCRIPTION
**Pull request by Remi-Gau**
Use pyproject.toml for config and package metadata.

### CODE REVIEW

File: tools/schemacode/pyproject.toml [hunk 1]
```diff
@@ -1,11 +1,65 @@
...
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
+    "Programming Language :: Python :: 3.13"
+]
+dependencies = [
+    "click",
+    "pyyaml",
+    "importlib_resources; python_version < '3.9'",
```
Comment (Remi-Gau) [inline tools/schemacode/pyproject.toml:21]: I guess this can go now, no?

File: tools/schemacode/pyproject.toml [hunk 2]
```diff
@@ -1,11 +1,65 @@
...
+    "jsonschema"
+]
+description = " Python tools for working with the BIDS schema."
+license = {text = "MIT"}
+maintainers = [{name = "bids-standard developers", email = " bids.maintenance@gmail.com"}]
+name = "bidsschematools"
+readme = "README.md"
+requires-python = ">=3.9"
+version = "0.11.4-dev"  # file:bidsschematools/data/schema/SCHEMA_VERSION
```
Comment (Remi-Gau) [inline tools/schemacode/pyproject.toml:30]: hard coded for now until I find how to read version from file

File: tools/schemacode/pyproject.toml [hunk 3]
```diff
@@ -1,11 +1,64 @@
 [build-system]
-requires = ["setuptools"]
-build-backend = "setuptools.build_meta"
+build-backend = "hatchling.build"
+requires = ["hatchling"]
+
+[project]
+authors = [{name = "Nilearn developers"}]
```
Comment (Remi-Gau) [inline tools/schemacode/pyproject.toml:6]: copy pasta...

File: tools/schemacode/pyproject.toml (hunk 2 above)
Comment (effigies) [reply tools/schemacode/pyproject.toml:29]:
```suggestion
dynamic = ["version"]
```

File: tools/schemacode/pyproject.toml [hunk 4]
```diff
@@ -1,11 +1,64 @@
...
+    "flake8-isort",
+    "pytest>6",
+    "pytest-cov"
+]
+
+[project.scripts]
+bst = "bidsschematools.__main__:cli"
+
+[project.urls]
+Homepage = "https://github.com/bids-standard/bids-specification"
```
Comment (effigies) [inline tools/schemacode/pyproject.toml:60]:
```suggestion

[tool.hatch.version]
path = "bidsschematools/data/schema/SCHEMA_VERSION
pattern = "(?P<version>.*)"
```

**Review by effigies — COMMENTED** [review]
I think...

**Comment by effigies** [comment]
I also think it's okay to keep using `setuptools`, if that's easiest.

**Comment by Remi-Gau** [comment]
> I also think it's okay to keep using `setuptools`, if that's easiest.

tried and it was a bit easier

**Comment by Remi-Gau** [comment]
Thanks https://github.com/bids-standard/bids-specification/pull/1948/commits/bc83dcbddd9b4d5a74074f6d24ebb7459da4a053 @effigies 

would have taken me a while to find it

**Comment by Remi-Gau** [comment]
This is good to go for me

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -61,6 +62,11 @@ repos:
       - id: yamllint
         args: [-f=standard, -c=.yamllint.yml]
         files: src/schema/.*/.*\.yaml
+  - repo: https://github.com/macisamuele/language-formatters-pre-commit-hooks
+    rev: v2.14.0
+    hooks:
+      - id: pretty-format-toml
+        args: [--autofix, --indent, '4']
```
Comment (effigies) [inline .pre-commit-config.yaml:69]: I don't think sorting leads to a very readable `pyproject.toml`, where I think there are some rough importance levels that are good to indicate (e.g. "name" should be the first thing in project metadata).

```suggestion
        args: [--autofix, --indent, '4', --no-sort]
```

File: tools/schemacode/pyproject.toml [hunk 6]
```diff
@@ -1,11 +1,64 @@
...
+    "pyyaml",
+    "jsonschema"
+]
+description = " Python tools for working with the BIDS schema."
+dynamic = ["version"]
+license = {text = "MIT"}
+maintainers = [{name = "bids-standard developers", email = " bids.maintenance@gmail.com"}]
+name = "bidsschematools"
+readme = "README.md"
+requires-python = ">=3.9"
```
Comment (effigies) [inline tools/schemacode/pyproject.toml:29]: Pure reordering:

```suggestion
name = "bidsschematools"
description = " Python tools for working with the BIDS schema."
authors = [{name = "bids-standard developers"}]
maintainers = [{name = "bids-standard developers", email = " bids.maintenance@gmail.com"}]
license = {text = "MIT"}
readme = "README.md"
requires-python = ">=3.9"
dependencies = [
    "click",
    "pyyaml",
    "jsonschema"
]
classifiers = [
    "Development Status :: 4 - Beta",
    "License :: OSI Approved :: MIT License",
    "Intended Audience :: Science/Research",
    "Topic :: Scientific/Engineering :: Information Analysis",
    "Programming Language :: Python :: 3.9",
    "Programming Language :: Python :: 3.10",
    "Programming Language :: Python :: 3.11",
    "Programming Language :: Python :: 3.12",
    "Programming Language :: Python :: 3.13"
]
dynamic = ["version"]
```

File: tools/schemacode/pyproject.toml [hunk 7]
```diff
@@ -1,11 +1,64 @@
...
+]
+tests = [
+    "codecov",
+    "coverage[toml]",
+    "flake8",
+    "flake8-black",
+    "flake8-isort",
+    "pytest>6",
+    "pytest-cov"
+]
```
Comment (effigies) [inline tools/schemacode/pyproject.toml:53]: Ordering: Optional functionality, tests, docs, all. Also, tests will fail if we don't include optional functionality, so adding that.

```suggestion
expressions = ["pyparsing"]
render = [
    "tabulate",
    "pandas",
    "markdown-it-py"
]
tests = [
    "bidsschematools[expressions,render]",
    "codecov",
    "coverage[toml]",
    "flake8",
    "flake8-black",
    "flake8-isort",
    "pytest>6",
    "pytest-cov"
]
doc = [
    "sphinx>=1.5.3",
    "sphinx_rtd_theme"
]
all = [
    "bidsschematools[doc,tests]"
]
```

File: tools/schemacode/pyproject.toml [hunk 8]
```diff
@@ -1,11 +1,64 @@
...
+bst = "bidsschematools.__main__:cli"
+
+[project.urls]
+Homepage = "https://github.com/bids-standard/bids-specification"
 
 [tool.black]
-line-length = 99
-target-version = ['py39']
-include = '\.pyi?$'
 exclude = '''
```
Comment (effigies) [inline tools/schemacode/pyproject.toml:62]:
```suggestion
line-length = 99
include = '\.pyi?$'
exclude = '''
```

File: tools/schemacode/pyproject.toml [hunk 9]
```diff
@@ -20,19 +73,43 @@ exclude = '''
   )/
 )
 '''
+include = '\.pyi?$'
+line-length = 99
+target-version = ['py39']
```
Comment (effigies) [inline tools/schemacode/pyproject.toml:78]:
```suggestion
```

The `target-version` is not necessary if `project.requires-python` is set.

File: tools/schemacode/pyproject.toml [hunk 10]
```diff
@@ -20,19 +73,43 @@ exclude = '''
...
+    "data/metaschema.json",
+    "data/schema/BIDS_VERSION",
+    "data/schema/SCHEMA_VERSIO",
+    "data/schema/**/*.yaml",
+    "tests/data/**/*",
+    "tests/data/**/.bidsignore"
 ]
+
+[tool.setuptools.packages.find]
+include = ["bidsschematools*"]
```
Comment (effigies) [inline tools/schemacode/pyproject.toml:115]: IMO packager config should come immediately after the `[project]` tables.

**Review by effigies — COMMENTED** [review]
Small nits.

File: .pre-commit-config.yaml (hunk 5 above)
Comment (Remi-Gau) [reply .pre-commit-config.yaml:69]: good with me
defo not a hill I want to die on

File: tools/schemacode/pyproject.toml (hunk 9 above)
Comment (Remi-Gau) [reply tools/schemacode/pyproject.toml:78]: ah yeah true

File: .pre-commit-config.yaml (hunk 5 above)
Comment (Remi-Gau) [reply .pre-commit-config.yaml:69]: forgot to commit this one

**Pull request merged**